main.go

```diff
@@ -165,13 +163,16 @@ func (s *prometheusServer) Start(ctx context.Context) error {
 
 	server.TLSConfig = s.getPrometheusTLSConfig(ctx, certWatcher)
 
-	if err := server.ListenAndServeTLS(s.certPath, s.keyPath); err != nil && !errors.Is(err, http.ErrServerClosed) {
+	if err = server.ListenAndServeTLS(s.certPath, s.keyPath); err != nil && !errors.Is(err, http.ErrServerClosed) {
 		setupLog.Error(err, "Failed to start Prometheus metrics endpoint server")
-		return err
+		cancel()
 	}
 
 	<-idleConnsClosed
```

Comment:
It would be better to use `sync.WaitGroup` here to wait for both of the goroutines to finish before the function returns.